File: examples/functions/python3/simple/harness.py
#! /usr/bin/env python3
# Step 1 - Package this in a linux container image.  Python runs fine in Linux.
# Note:
# This harness exits with a success code even in cases of error as  
# kubernetes jobs will re-run until it sees a successful exit code.
# Errors are populated through the backend API back to the backend API client.
import os
import grpc
import mmf 
import api.protobuf_spec.messages_pb2 as mmlogic
import api.protobuf_spec.mmlogic_pb2_grpc as mmlogic_pb2_grpc
from google.protobuf.json_format import Parse
import ujson  as json
import pprint as pp

from timeit import default_timer as timer

# Load config file
cfg = ''
with open("matchmaker_config.json") as f:
    cfg = json.loads(f.read())

# Step 2 - Talk to Redis.  This example uses the MM Logic API in OM to read/write to/from redis.
# Establish grpc channel and make the API client stub
api_conn_info = "%s:%d" % (cfg['api']['mmlogic']['hostname'], cfg['api']['mmlogic']['port'])
with  grpc.insecure_channel(api_conn_info) as channel:
    mmlogic_api = mmlogic_pb2_grpc.MmLogicStub(channel)

    # Step 3 - Read the profile written to the Backend API.
    # Get profile from redis
    profile_pb = mmlogic_api.GetProfile(mmlogic.MatchObject(id=os.environ["MMF_PROFILE_ID"]))
    pp.pprint(profile_pb) #DEBUG
    profile_dict = json.loads(profile_pb.properties)

    # Step 4 - Select the player data from Redis that we want for our matchmaking logic.
    # Embedded in this profile are JSON representations of the filters for each player pool.
    # JsonFilterSet() is able to read those directly.  No need to marhal that
    # JSON into the protobuf message format.
    player_pools = dict()       # holds dictionary version of the pools. 
    for empty_pool in profile_pb.pools:
        # Dict to hold value-sorted field dictionary for easy retreival of players by value
        player_pools[empty_pool.name] = dict() 
        print("Retrieving pool '%s'" % empty_pool.name, end='')

        # DEBUG: Print how long the filtering takes
        if cfg['debug']:
            start = timer()

        # Pool filter results are streamed in chunks as they can be too large to send
        # in one grpc message.  Loop to get them all.
        for partial_results in mmlogic_api.GetPlayerPool(empty_pool):
            empty_pool.stats.count = partial_results.stats.count
            empty_pool.stats.elapsed = partial_results.stats.elapsed
            print(".", end='')
            try:
                for player in partial_results.roster.players:
                    if not player.id in player_pools[empty_pool.name]:
                        player_pools[empty_pool.name][player.id] = dict()
                    for prop in player.properties:
                        player_pools[empty_pool.name][player.id][prop.name] = prop.value
            except Exception as err:
                print("Error encountered: %s" % err) 
        if cfg['debug']:
            print("\n'%s': count %06d | elapsed %0.3f" % (empty_pool.name, len(player_pools[empty_pool.name]),timer() - start))

    #################################################################
    # Step 5 - Run custom matchmaking logic to try to find a match
    # This is in the file mmf.py
    results = mmf.makeMatches(profile_dict, player_pools)
    #################################################################

    # DEBUG
    if cfg['debug']:
        print("======== match_properties")
        pp.pprint(results) 

    # Generate a MatchObject message to write to state storage with the results in it.
    # This looks odd but it's how you assign to repeated protobuf fields.
    # https://developers.google.com/protocol-buffers/docs/reference/python-generated#repeated-fields
    match_properties = json.dumps(results)
    mo = mmlogic.MatchObject(id = os.environ["MMF_PROPOSAL_ID"], properties = match_properties)
    mo.pools.extend(profile_pb.pools[:])

    # Access the rosters in dict form within the properties json.
    # It is stored at the key specified in the config file.
    rosters_dict = results
    for partial_key in cfg['jsonkeys']['rosters'].split('.'):
        rosters_dict = rosters_dict.get(partial_key, {})

    # Unmarshal the rosters into the MatchObject 
    for roster in rosters_dict:
        mo.rosters.extend([Parse(json.dumps(roster), mmlogic.Roster(), ignore_unknown_fields=True)])

    #DEBUG: writing to error key prevents evalutor run 
    if cfg['debug']
        mo.id = os.environ["MMF_ERROR_ID"] 
        mo.error = "skip evaluator"
        print("======== MMF results:") 
        pp.pprint(mo)

    # Step 6 - Write the outcome of the matchmaking logic back to state storage.
    # Step 7 - Remove the selected players from consideration by other MMFs.
    # CreateProposal does both of these for you, and some other items as well.
    success = mmlogic_api.CreateProposal(mo)
    print("======== MMF write to state storage:  %s" % success) 


#    # [OPTIONAL] Step 8 - Export stats about this run.
#    # TODO 

File: examples/functions/golang/serving-mmlogic-simple/apisrv/harness.go
/*
 Note:
 This is a harness for rigging up custom matchmaking logic to a gRPC API
 endpoint, so that other services can make a simple gRPC request to kick off
 the logic.  The harness is optional, but tries to handle many of the
 boilerplate steps that matchmaking functions are required to execute (outlined
 in the <REPO_ROOT>/README.md file).  If you don't want to use the MMLogic API
 and would rather your MMF talk directly to Redis to do something else, feel
 free to not use this harness at all and instead write your own mmfRun()
 function with the specified function signature.  Errors are populated through
 the backend API back to the backend API client */
package apisrv

import (
	"context"
	"io"
	"time"

	"github.com/golang/protobuf/proto"
	log "github.com/sirupsen/logrus"

	"github.com/GoogleCloudPlatform/open-match/internal/logging"
	api "github.com/GoogleCloudPlatform/open-match/internal/pb"
	"github.com/spf13/viper"

	"go.opencensus.io/stats"
)

var (
	// Logrus structured logging setup
	mmfLogFields = log.Fields{
		"app":       "openmatch",
		"component": "function_mmlogic_harness",
	}
	mmfLog = log.WithFields(mmfLogFields)
)

// Step 2 - Talk to Redis.
// This example uses the MM Logic API in OM to read/write to/from redis.

// mmfRun is used to kick off the served MMF.
func mmfRun(ctx context.Context, fnArgs *api.Arguments, cfg *viper.Viper, mmlogic api.MmLogicClient) error {

	// Configure open match logging defaults
	logging.ConfigureLogging(cfg)

	// Get a cancel-able context
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	// Prepare to measure how long this takes
	var start time.Time
	runtime := time.Now()
	defer stats.Record(ctx, FnLatencySecs.M(time.Since(runtime).Seconds()))

	// Step 3 - Read the profile written to the Backend API
	profile, err := mmlogic.GetProfile(ctx, &api.MatchObject{Id: fnArgs.Request.ProfileId})
	if err != nil {
		mmfLog.Error(err)
		stats.Record(ctx, FnFailures.M(1))
		return err
	}
	if cfg.IsSet("debug") && cfg.GetBool("debug") {
		mmfLog.Debug("Profile: ", profile)
	}

	// Step 4 - Select the player data from Redis that we want for our matchmaking logic.
	playerPools := make([]*api.PlayerPool, len(profile.Pools))
	numPlayers := 0
	for index, emptyPool := range profile.Pools {
		poolLog := mmfLog.WithFields(log.Fields{"poolName": emptyPool.Name})
		playerPools[index] = proto.Clone(emptyPool).(*api.PlayerPool)
		playerPools[index].Roster = &api.Roster{Players: []*api.Player{}}
		poolLog.Info("Retrieving pool")
		// Taking out pool name for cardinality when procedurally generating
		// names. If you aren't procedurally generating names, having this
		// attached will let you get a lot of interesting metrics aggregated by
		// pool.
		//poolCtx, _ := tag.New(ctx, tag.Insert(KeyPoolName, "aggregate"))
		poolCtx := ctx

		// DEBUG: Print how long the filtering takes
		if cfg.IsSet("debug") && cfg.GetBool("debug") {
			start = time.Now()
		}

		// Pool filter results are streamed in chunks as they can be too large
		// to send in one grpc message.  Loop to get them all.
		stream, err := mmlogic.GetPlayerPool(ctx, emptyPool)
		if err != nil {
			poolLog.Error(err)
			stats.Record(ctx, FnFailures.M(1))
			return err
		}
		for {
			partialResults, err := stream.Recv()
			if err == io.EOF {
				// Break when all results are received
				break
			}
			if err != nil {
				poolLog.Error(err)
				stats.Record(ctx, FnFailures.M(1))
				return err
			}

			// Update stats with the latest results
			emptyPool.Stats = partialResults.Stats

			// Put players into the Pool's Roster with the attributes we
			// already retrieved through our filter queries..
			if partialResults.Roster != nil && len(partialResults.Roster.Players) > 0 {
				for _, player := range partialResults.Roster.Players {
					playerPools[index].Roster.Players = append(playerPools[index].Roster.Players, proto.Clone(player).(*api.Player))

					// Count players retrieved, used to short circuit logic if no players matched the filters.
					numPlayers++
				}
			}

			if cfg.IsSet("debug") && cfg.GetBool("debug") {
				poolLog.WithFields(log.Fields{
					"elapsed": time.Now().Sub(start).Seconds(),
					"count":   len(playerPools[index].Roster.Players),
				}).Debug("retrieval stats")
			}
		}
		if emptyPool.Stats != nil {
			// Emit metrics for MMFs to Open Census.
			stats.Record(poolCtx, FnPoolPlayersRetreivedTotal.M(emptyPool.Stats.Count))
			stats.Record(poolCtx, FnPoolElapsedSeconds.M(emptyPool.Stats.Elapsed))
			log.Infof(" Pool latency: %0.2v", emptyPool.Stats.Elapsed)
		}
	}

	// Generate a MatchObject message to write to state storage with the
	// results in it.  By default, assume we weren't successful (set ID to
	// ResultID, which skips the evaluator when outputting errors) until proven
	// otherwise.
	mo := &api.MatchObject{
		Id:         fnArgs.Request.ResultId,
		Properties: profile.Properties,
		Pools:      profile.Pools,
	}

	// Return error when there are no players in the pools
	if numPlayers == 0 {
		if cfg.IsSet("debug") && cfg.GetBool("debug") {
			mmfLog.Info("All player pools are empty, writing directly to ResultID to skip the evaluator")
		}
		// Writing to the error key skips the evaluator
		mo.Error = "insufficient players"
		stats.Record(ctx, FnFailures.M(1))
	} else {

		////////////////////////////////////////////////////////////////
		// Step 5 - Run custom matchmaking logic to try to find a match
		// This is in the file mmf.go.  Customize with your own match logic!
		results, rosters, err := makeMatches(ctx, profile.Properties, profile.Rosters, playerPools)
		////////////////////////////////////////////////////////////////

		if err != nil {
			mmfLog.WithFields(log.Fields{
				"error": err.Error(),
				"id":    fnArgs.Request.ResultId,
			}).Error("MMF returned an unrecoverable error")
			stats.Record(ctx, FnFailures.M(1))
			mo.Error = err.Error()
		} else {
			// No error!
			// Prepare the proposal match object.
			mo.Id = fnArgs.Request.ProposalId
			mo.Properties = results
			mo.Rosters = rosters
		}

	}

	// DEBUG
	if cfg.IsSet("debug") && cfg.GetBool("debug") {
		mmfLog.Debug("Output MatchObject:", mo)
	}

	// Step 6 - Write the outcome of the matchmaking logic back to state storage.
	// Step 7 - Remove the selected players from consideration by other MMFs.
	// CreateProposal does both of these for you, and some other tasks as well.
	success, err := mmlogic.CreateProposal(ctx, mo)
	if err != nil {
		mmfLog.Error(err)
		stats.Record(ctx, FnFailures.M(1))
	} else {
		mmfLog.WithFields(log.Fields{"id": fnArgs.Request.ProposalId, "success": success.Success}).Info("MMF write to state storage")
	}

	// Step 8 - Export stats about this run.
	stats.Record(ctx, FnRequests.M(1))
	return nil
}
